File: sample-objc/ARGEARSample/UI/UIComponent/Main/Bottom/Beauty/BeautyView.m
//
//  BeautyView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "BeautyView.h"
#import "UIImage+Color.h"
#import "BulgeManager.h"
#import "StickerManager.h"

@implementation BeautyView

#define kBeautyViewCompareIconName @"icCompare"
#define kBeautyViewRefreshIconName @"icRefresh"

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (void)awakeFromNib {
    [super awakeFromNib];
    
    _gradient = [CAGradientLayer layer];
    [_gradient setFrame:[self bounds]];
    [_gradient setColors:@[(id)_topGradientColor.CGColor, (id)_bottomGradientColor.CGColor]];
    
    UIImage *sliderThumb = [UIImage imageNamed:@"bttnControl.png"];
    [_slider setThumbImage:sliderThumb forState:UIControlStateNormal];
    [_slider setThumbImage:sliderThumb forState:UIControlStateHighlighted];
    [_slider setValue:[[BeautyManager shared] getBeautyValue:0]];

    UILongPressGestureRecognizer *longPressGesture = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(compareButtonLongPressAction:)];
    [_compareButton addGestureRecognizer:longPressGesture];
    
    [self setButtons:[self tag]];
    [self addObservers];
}

- (void)dealloc {
    [self removeObservers];
}

- (void)open {
    [self setHidden:NO];
    
    [[BulgeManager shared] off];
    [[StickerManager shared] clearSticker];
}

- (void)close {
    [self setHidden:YES];
}

- (void)setRatio:(ARGMediaRatio)ratio {
    
    [self setTag:ratio];
    
    if (ratio == ARGMediaRatio_16x9) {

        [[self layer] insertSublayer:_gradient atIndex:0];
        [self setBackgroundColor:[UIColor clearColor]];
    } else {
        
        [_gradient removeFromSuperlayer];
        [self setBackgroundColor:[UIColor whiteColor]];
    }
    
    [self setButtons:ratio];
    
    [_beautyCollectionView setRatio:ratio];
}

- (void)setButtons:(ARGMediaRatio)ratio {

    NSMutableString *compareButtonImageString = [NSMutableString stringWithString:kBeautyViewCompareIconName];
    NSMutableString *refreshButtonImageString = [NSMutableString stringWithString:kBeautyViewRefreshIconName];
    
    if (ratio == ARGMediaRatio_16x9) {
        
        [compareButtonImageString appendString:@"White"];
        [refreshButtonImageString appendString:@"White"];
    } else {
        
        [compareButtonImageString appendString:@"Black"];
        [refreshButtonImageString appendString:@"Black"];
    }
    
    [_compareButton setImage:[UIImage imageNamed:compareButtonImageString] forState:UIControlStateNormal];
    [_compareButton setImage:[[UIImage imageNamed:compareButtonImageString] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];

    [_resetButton setImage:[UIImage imageNamed:refreshButtonImageString] forState:UIControlStateNormal];
    [_resetButton setImage:[[UIImage imageNamed:refreshButtonImageString] imageWithAlphaComponent:BUTTON_HIGHLIGHTED_ALPHA] forState:UIControlStateHighlighted];
}

- (void)addObservers {
    [_beautyCollectionView addObserver:self forKeyPath:@"selectedIndexPath" options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld context:nil];
}

- (void)removeObservers {
    [_beautyCollectionView removeObserver:self forKeyPath:@"selectedIndexPath" context:nil];
}

- (void)compareButtonLongPressAction:(UILongPressGestureRecognizer*)gesture {
    if ([gesture state] == UIGestureRecognizerStateBegan) {
        [[BeautyManager shared] off];
    } else if ([gesture state] == UIGestureRecognizerStateEnded) {
        [[BeautyManager shared] on];
    }
}

- (IBAction)resetButtonAction:(UIButton *)sender {
    
    [[BeautyManager shared] setDefault];
    
    NSIndexPath *selectedIndexPath = [_beautyCollectionView selectedIndexPath];
    CGFloat beautyValue = [[BeautyManager shared] getBeautyValue:[selectedIndexPath row]];
    
    [_slider setValue:beautyValue];
}

#pragma mark - UISlider
- (IBAction)sliderValueChanged:(UISlider *)sender {
    
    [[BeautyManager shared] setBeauty:[sender tag] value:[sender value]];
}

# pragma mark - Observers
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    
    if ([object isKindOfClass:[BeautyCollectionView class]]) {
        
        if ([keyPath isEqualToString:@"selectedIndexPath"]) {
            NSIndexPath *indexPath = [change objectForKey:@"new"];
            
            CGFloat beautyValue = [[BeautyManager shared] getBeautyValue:[indexPath row]];
            [_slider setTag:[indexPath row]];
            [_slider setValue:beautyValue];
        }
    }
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Main/Bottom/Beauty/BeautyView.swift
//
//  BeautyView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/21.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import ARGear

class BeautyView: ARGBottomFunctionBaseView {
    
    let kBeautyViewCompareIconName = "icCompare"
    let kBeautyViewResetIconName = "icRefresh"
    
    @IBOutlet weak var slider: UISlider!
    
    @IBOutlet weak var compareButton: UIButton!
    @IBOutlet weak var resetButton: UIButton!
    
    @IBOutlet weak var beautyCollectionView: BeautyCollectionView!
    
    private var observers = [NSKeyValueObservation]()
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        let sliderThumb = UIImage(named: "bttnControl.png")
        self.slider.setThumbImage(sliderThumb, for: .normal)
        self.slider.setThumbImage(sliderThumb, for: .highlighted)

        self.beautyCollectionView.selectedIndexPath = IndexPath(row: 0, section: 0)
        
        let longPressGesture = UILongPressGestureRecognizer(target: self, action: #selector(compareButtonLongPressAction(_:)))
        self.compareButton.addGestureRecognizer(longPressGesture)
        
        self.addObservers()
    }
    
    deinit {
        self.removeObservers()
    }
    
    override func open() {
        super.open()

        var beautyIndex = 0
        if let selectedBeauty = self.beautyCollectionView.selectedIndexPath {
            beautyIndex = selectedBeauty.row
        }
        self.slider.value = BeautyManager.shared.getBeautyValue(type: ARGContentItemBeauty(rawValue: beautyIndex)!)
        
        BulgeManager.shared.off()
        ContentManager.shared.clearContent()
    }
    
    override func setRatio(_ ratio: ARGMediaRatio) {
        super.setRatio(ratio)
        
        self.setButtons(ratio: ratio)
        self.beautyCollectionView.setRatio(ratio)
    }
    
    func setButtons(ratio: ARGMediaRatio) {
        
        var compareButtonImageName = self.kBeautyViewCompareIconName
        var resetButtonImageName = self.kBeautyViewResetIconName
        
        var imageColorString = ""
        if ratio == ._16x9 {
            imageColorString = "White"
        } else {
            imageColorString = "Black"
        }
        compareButtonImageName.append(imageColorString)
        resetButtonImageName.append(imageColorString)

        self.compareButton.setImage(UIImage(named: compareButtonImageName), for: .normal)
        self.compareButton.setImage(UIImage(named: compareButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)

        self.resetButton.setImage(UIImage(named: resetButtonImageName), for: .normal)
        self.resetButton.setImage(UIImage(named: resetButtonImageName)?.withAlpha(BUTTON_HIGHLIGHTED_ALPHA), for: .highlighted)
    }
    
    @objc
    func compareButtonLongPressAction(_ gesture: UILongPressGestureRecognizer) {
        if gesture.state == .began {
            BeautyManager.shared.off()
        } else if gesture.state == .ended {
            BeautyManager.shared.on()
        }
    }
    
    @IBAction func resetButtonAction(_ sender: UIButton) {
        BeautyManager.shared.setDefault()
        
        if let selectedIndexPath = beautyCollectionView.selectedIndexPath {
            let beautyValue = BeautyManager.shared.getBeautyValue(type: ARGContentItemBeauty(rawValue: selectedIndexPath.row) ?? .vline)
            
            slider.setValue(beautyValue, animated: false)
        }
    }
    
    @IBAction func sliderValueChanged(_ sender: UISlider) {
        BeautyManager.shared.setBeauty(type: ARGContentItemBeauty(rawValue: sender.tag)!, value: sender.value)
    }
    
    func addObservers() {
        self.observers.append(
            self.beautyCollectionView.observe(\.selectedIndexPath, options: .new) { [weak self] _, change in
                guard let self = self else { return }

                if let indexPath = change.newValue as? IndexPath {
                    let beautyValue = BeautyManager.shared.getBeautyValue(type: ARGContentItemBeauty(rawValue: indexPath.row)!)
                    self.slider.tag = indexPath.row
                    self.slider.value = beautyValue
                }
            }
        )
    }
    
    func removeObservers() {
        self.observers.removeAll()
    }
}
